modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsRequestCollection.java

```diff
@@ -0,0 +1,381 @@
+
+    /**
+     * Constructor.
+     *
+     * @param schemaMgr Schema manager.
+     */
+    public IgniteStatisticsRequestCollection(SchemaManager schemaMgr) {
+        this.schemaMgr = schemaMgr;
+    }
+
+    /**
+     * Add new statistics collection status.
+     *
+     * @param id Id to save status by.
+     * @param status status to add.
+     */
+    public void addActiveCollectionStatus(UUID id, StatCollectionStatus status) {
+        currColls.put(id, status);
+    }
+
+    /**
+     * Update status of statistics gathering task.
+     *
+     * @param id Statistics collection task id.
+     * @param transformation Transformation to apply, if it returns {@code null} - status will be removed.
+     */
+    public void updateCollection(UUID id, Function<StatCollectionStatus, StatCollectionStatus> transformation) {
+        currColls.compute(id, (k, v) -> transformation.apply(v));
+    }
+
+    /**
+     * Get collection status.
+     *
+     * @param id Id to get status by.
+     * @return Collection status.
+     */
+    public StatCollectionStatus getCollection(UUID id) {
+        return currColls.get(id);
+    }
+
+
+    /**
+     * Extract groups of stats keys.
+     *
+     * @param keys Statistics key to extract groups from.
+     * @return Map of <group ids></group> to <collection of keys in groups>
+     * @throws IgniteCheckedException In case of lack some of specified objects.
+     */
+    protected Map<CacheGroupContext, Collection<StatsKeyMessage>> extractGroups(Collection<StatsKeyMessage> keys)
+            throws IgniteCheckedException {
+        Map<CacheGroupContext, Collection<StatsKeyMessage>> res = new HashMap<>(keys.size());
+        for (StatsKeyMessage key : keys) {
+            GridH2Table tbl = schemaMgr.dataTable(key.schema(), key.obj());
+            if (tbl == null)
+                throw new IgniteCheckedException(String.format("Can't find object %s.%s", key.schema(), key.obj()));
+
+            res.compute(tbl.cacheContext().group(), (k, v) -> {
+                if (v == null)
+                    v = new ArrayList<>();
+
+                v.add(key);
+
+                return v;
+            });
+        }
+        return res;
+    }
+
+    /**
+     * Get map cluster node to it's partitions for the specified cache group.
+     *
+     * @param grp Cache group context to get partition information by.
+     * @param partIds Partition to collect information by
+     * @return Map nodeId to array of partitions, related to node.
+     */
+    public static Map<UUID, int[]> nodePartitions(CacheGroupContext grp, Collection<Integer> partIds)
+            throws IgniteCheckedException {
+        AffinityTopologyVersion grpTopVer = grp.shared().exchange().readyAffinityVersion();
+        // На какой ноде на какой таблице какие партиции собирать
+        List<List<ClusterNode>> assignments = grp.affinity().assignments(grpTopVer);
+
+        Map<UUID, List<Integer>> res = new HashMap<>();
+        if (partIds == null)
```

Comment:
this could be simplified by generating partition ids from assignments

Reply:
Sometimes we need primary, sometimes - backups, sometimes - only mapping for some specied partitions... now this code located in IgniteStatisticsHelper.nodePartitions. How I can simplify it?

Reply:
for example like this:

`

        if (partIds == null)
            partIds = IntStream.range(0, assignments.size()).boxed().collect(Collectors.toList());

        for (Integer partId : partIds)
            fillPartition(res, assignments, partId, isPrimary);`

Reply:
fixed